File: 600/695 - Max Area of Island.py
class Solution:
    directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]
    
    def search(self, grid, mx, my, x, y, visited):
        if visited[x][y]:
            return 0

        visited[x][y] = True
        if grid[x][y] == 0:
            return 0

        ret = 1
        for d in self.directions:
            nx, ny = x + d[0], y + d[1]
            if 0 <= nx < mx and 0 <= ny < my:
                ret += self.search(grid, mx, my, nx, ny, visited)

        return ret
    
    def maxAreaOfIsland(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        mx, my = len(grid), len(grid[0])
        visited = [[False for _ in range(my)] for _ in range(mx)]
        ret = 0
        for i in range(mx):
            for j in range(my):
                ret = max(ret, self.search(grid, mx, my, i, j, visited))
        return ret

File: csharp/695.max-area-of-island.cs
/*
 * @lc app=leetcode id=695 lang=csharp
 *
 * [695] Max Area of Island
 */

using System.Collections.Generic;

// @lc code=start
public class Solution {
    
    private static readonly int[][] Dirs = new int[][] {
        new int[] { 0, -1, },
        new int[] { 0, 1, },
        new int[] { -1, 0, },
        new int[] { 1, 0, },
    };

    public int MaxAreaOfIsland(int[][] grid) {
        int m = grid.Length, n = grid[0].Length, ret = 0;
        Queue<(int, int)> queue = new Queue<(int, int)>();
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (grid[i][j] != 1) {
                    continue;
                }
                int curr = 0;
                grid[i][j] = -1;
                queue.Enqueue((i, j));
                while (queue.Count > 0) {
                    ++curr;
                    var l = queue.Dequeue();
                    foreach (var dir in Dirs) {
                        int nr = l.Item1 + dir[0], nc = l.Item2 + dir[1];
                        if (nr < 0 || nr >= m || nc < 0 || nc >= n || grid[nr][nc] != 1) {
                            continue;
                        }
                        grid[nr][nc] = -1;
                        queue.Enqueue((nr, nc));
                    }
                }
                ret = Math.Max(ret, curr);
            }
        }
        return ret;
    }
}
// @lc code=end

